jax/_src/debugging.py

```diff
@@ -0,0 +1,147 @@
+
+# TODO(sharadmv): remove this global keepalive list in favor of attaching
+# keepalives to the module context.
+_keepalives = []
+
+def _ordered_effect_lowering(ctx, token, *args, **params):
+  avals_in = [core.abstract_token, *ctx.avals_in]
+  avals_out = [core.abstract_token, *ctx.avals_out]
+  args = (token, *args)
+  def _callback(token, *flat_args):
+    out = debug_callback_p.impl(*flat_args, **params)
+    return (token, *out)
+  (token, *result), keepalive = mlir.emit_python_callback(
+      ctx.module_context.platform, _callback, list(args), avals_in, avals_out,
+      True)
+  return result, keepalive, token
+
+def debug_callback_lowering(ctx, *args, effect, callback, **params):
+  if effect in core.ordered_effects:
+    token = ctx.tokens_in.get(effect)[0]
+    result, keepalive, token = _ordered_effect_lowering(ctx, token,
+        *args, effect=effect, callback=callback, **params)
+    ctx.set_tokens_out(mlir.TokenSet({effect: (token,)}))
+  else:
+    def _callback(*flat_args):
+      return tuple(debug_callback_p.impl(
+        *flat_args, effect=effect, callback=callback, **params))
+    result, keepalive = mlir.emit_python_callback(ctx.module_context.platform,
+      _callback, list(args), ctx.avals_in, ctx.avals_out,  True)
+  _keepalives.append(keepalive)
+  return result
+mlir.register_lowering(debug_callback_p, debug_callback_lowering,
+                       platform="cpu")
+
+def debug_callback(callback: Callable[..., Any],
+                   effect: DebugEffect, result_shapes: Any, *args, **kwargs):
+  """Calls a stageable Python callback.
+
+  `debug_callback` enables you to pass in a Python function that can be called
+  inside of a staged JAX program.
+
+  Args:
+    callback: A Python callable.
+    effect: A `DebugEffect`.
+    result_shapes: A Pytree of `jax.ShapeDtypeStruct`-like objects that match
+      output of `callback`.
+      *args: The positional arguments to the callback.
+      **kwargs: The positional arguments to the callback.
+  Returns:
+    The results of `callback` applied to `*args` and `**kwargs`.
+  """
+  if not isinstance(effect, DebugEffect):
+    raise ValueError("Can only use `DebugEffect` effects in `debug_callback`")
+  flat_args, in_tree = tree_util.tree_flatten((args, kwargs))
+  out_avals = tree_util.tree_map(lambda x: core.ShapedArray(x.shape, x.dtype),
+                                 result_shapes)
+  return debug_callback_p.bind(*flat_args, callback=callback, effect=effect,
+                               in_tree=in_tree, out_avals=out_avals)
+
+def _format_print_callback(fmt: str, *args, **kwargs):
+  sys.stdout.write(fmt.format(*args, **kwargs) + "\n")
+
+def debug_print(fmt: str, *args, ordered=False, **kwargs) -> None:
+  """Prints values and works in staged out JAX functions.
+
+  Args:
+    fmt: A format string, e.g. `"hello {x}"`, that will be used to format
+      input arguments.
+    *args: A list of positional JAX arguments.
```

Comment:
What's "JAX" mean here? Maybe arrays? Is there any difference with what's passed in `kwargs`? (I ask b/c the description of the kwargs doesn't mention "JAX".)

Reply:
Fixed, removed use of "JAX".